### DISCUSSION TITLE
Read Only and Init only properties doesn't map on Load

### DISCUSSION THREAD

**Discussion by Hrsrocha** [ravendb/ravendb#14717] [Q&A]
Hi,

I'm folowing the Domain Driven Design model, and i have some unexpected scenarios when i try to load some kind of properties.
Class with Backfield and readonly list properties:
```
    public class ClassWithBackFieldAndReadOnlyList
    {
        public string Id { get; set; }
        public string AnotherProperty { get; private set; }
        private List<FakeModel> _myList;
        public IReadOnlyList<FakeModel> MyList => _myList.AsReadOnly();

        public ClassWithBackFieldAndReadOnlyList(string anotherProperty)
        {
            AnotherProperty = anotherProperty;
            _myList = new List<FakeModel>();
        }

        public void AddFakeModelToList(FakeModel fakeModel)
        {
            _myList.Add(fakeModel);
        }
    }

    public class FakeModel
    {
        public string FakeProperty { get; set; }
    }
```
Class with init only property:
```
    public class ClassWithInitOnlyProperty
    {
        public string Id { get; set; }
        public DateTime CreatedDate { get; }

        public ClassWithInitOnlyProperty()
        {
            CreatedDate = DateTime.Now;
        }
    }
```
When i load this classes from ravendb, they have a null value.
How do i load this properties correclty? 
I made an example with my problem [My Example](https://github.com/Hrsrocha/RavenDBPropertyMap/blob/main/RavenDBPropertiesMapping.Tests/RavenDBMappingTests.cs)

**Comment by garayx** [comment]
Hi,

Thanks we will take a look on the failing tests, please follow https://issues.hibernatingrhinos.com/issue/RavenDB-19134/Read-Only-and-Init-only-properties-doesnt-map-on-Load for progress.

**Comment by shaharhikri** [comment]
Which version of ravendb are you working with?

**Reply by Hrsrocha** [reply]
I got the same issues with the versions:

- .Net 6.0
- RavenDB.Client 5.4.1
- RavenDB 5.3 And RavenDB 5.4

- .Net 2.2
- RavenDB.Client 5.2.0
- RavenDB 5.3 And RavenDB 5.4

**Comment by ayende** [comment]
You need to customize the serialization here, see:
https://www.newtonsoft.com/json/help/html/JsonConstructorAttribute.htm

https://stackoverflow.com/questions/64749277/in-c9-how-do-init-only-properties-differ-from-read-only-properties

**Reply by Hrsrocha** [reply]
I followed this approach, but i was looking for a better solution. In my classess i added private constructors with [JsonConstructor] attribute, it's ok, but with more complexity i think it will be harm to read some classes

**Comment by shaharhikri** [comment]
There's 3 ways to do that:
**1.** You can use this custom ```ContractResolver```:
```
        private class MyContractResolver : DefaultContractResolver
        {
            protected override JsonObjectContract CreateObjectContract(Type objectType)
            {
                var contract = base.CreateObjectContract(objectType);
                contract.MemberSerialization = MemberSerialization.Fields;
                return contract;
            }

            protected override List<MemberInfo> GetSerializableMembers(Type objectType)
            {
                var fields = objectType.GetFields(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly).Where(f => !f.IsNotSerialized);
                return fields.Cast<MemberInfo>().ToList();
            }

            protected override IList<JsonProperty> CreateProperties(Type type, MemberSerialization memberSerialization)
            {
                return base.CreateProperties(type, MemberSerialization.Fields);
            }
        }
```
When you create the ```DocumentStore``` just add ```MyContractResolver``` to its 'Conventions' like that (before you initialize the ```DocumentStore```):
```
var store = new DocumentStore()
{
    Urls = new string[] { "http://localhost:8080" },
    Database = "Test"
};

store.Conventions.Serialization= new NewtonsoftJsonSerializationConventions {JsonContractResolver = new MyContractResolver() }; // <=

store.Initialize();
```
.
.
.
**2.** You can add a private JsonConstructor to ```ClassWithInitOnlyProperty``` like this:
```
    public class ClassWithInitOnlyProperty
    {
        public string Id { get; set; }
        public DateTime CreatedDate { get; }

        public ClassWithInitOnlyProperty()
        {
            CreatedDate = DateTime.Now;
        }

        [JsonConstructor]
        private ClassWithInitOnlyProperty(DateTime createdDate) // <==
        {
            CreatedDate = createdDate;
         }
    }
```
.
.
.
**3.** You can add 'init' to the "CreatedDate" property:
```
    public class ClassWithInitOnlyProperty
    {
        public string Id { get; set; }
        public DateTime CreatedDate { get; init; } // <==

        public ClassWithInitOnlyProperty()
        {
            CreatedDate = DateTime.Now;
        }
   }
```

**Reply by shaharhikri** [reply]
@Hrsrocha
